src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2173,3 +2173,209 @@ Base.getindex(::AbstractDataFrame, ::Union{Symbol, Integer, AbstractString}) =
+```
+"""
+function insertcols!(df::AbstractDataFrame, col::ColumnIndex, name_cols::Pair{Symbol, <:Any}...;
+                     makeunique::Bool=false, copycols::Bool=true)
+    if !(df isa DataFrame || (df isa SubDataFrame && getfield(df, :colindex) isa Index))
+        throw(ArgumentError("insertcols! is only supported for DataFrame or" *
+                            "SubDataFrame created with `:` as column selector"))
+    end
+    col_ind = Int(col isa SymbolOrString ? columnindex(df, col) : col)
+    if !(0 < col_ind <= ncol(df) + 1)
+        throw(ArgumentError("attempt to insert a column to a data frame with " *
+                            "$(ncol(df)) columns at index $col_ind"))
+    end
+
+    if !makeunique
+        if !allunique(first.(name_cols))
+            throw(ArgumentError("Names of columns to be inserted into a data frame " *
+                                "must be unique when `makeunique=true`"))
+        end
+        for (n, _) in name_cols
+            if hasproperty(df, n)
+                throw(ArgumentError("Column $n is already present in the data frame " *
+                                    "which is not allowed when `makeunique=true`"))
+            end
+        end
+    end
+
+    if ncol(df) == 0 && df isa DataFrame
+        target_row_count = -1
+    else
+        target_row_count = nrow(df)
+    end
+
+    for (n, v) in name_cols
+        if v isa AbstractVector
+            if target_row_count == -1
+                target_row_count = length(v)
+            elseif length(v) != target_row_count
+                if target_row_count == nrow(df)
+                    throw(DimensionMismatch("length of new column $n which is " *
+                                            "$(length(v)) must match the number " *
+                                            "of rows in data frame ($(nrow(df)))"))
+                else
+                    throw(DimensionMismatch("all vectors passed to be inserted into " *
+                                            "a data frame must have the same length"))
+                end
+            end
+        elseif v isa AbstractArray && ndims(v) > 1
+            throw(ArgumentError("adding AbstractArray other than AbstractVector as " *
+                                "a column of a data frame is not allowed"))
+        end
+    end
+    if target_row_count == -1
+        target_row_count = 1
+    end
+
+    for (name, item) in name_cols
+        if !(item isa AbstractVector)
+            if item isa Union{AbstractArray{<:Any, 0}, Ref}
+                x = item[]
+                item_new = fill!(Tables.allocatecolumn(typeof(x), target_row_count), x)
+            else
+                @assert !(item isa AbstractArray)
+                item_new = fill!(Tables.allocatecolumn(typeof(item), target_row_count), item)
+            end
+        elseif item isa AbstractRange
+            item_new = collect(item)
+        elseif copycols
+            item_new = copy(item)
```

Comment:
These steps are not needed when `df isa SubDataFrame`, right?

Reply:
yes - copying is not needed then. I will fix this